src/doc/man/fedfs-create-junction.8
.\"@(#)fedfs-create-junction.8"
.\"
.\" @file doc/man/fedfs-create-junction.8
.\" @brief man page for fedfs-create-junction client command
.\"

.\"
.\" Copyright 2011 Oracle.  All rights reserved.
.\"
.\" This file is part of fedfs-utils.
.\"
.\" fedfs-utils is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License version 2.0 as
.\" published by the Free Software Foundation.
.\"
.\" fedfs-utils is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License version 2.0 for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" version 2.0 along with fedfs-utils.  If not, see:
.\"
.\"	http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt
.\"
.TH FEDFS-CREATE-JUNCTION 8 "2 July 2013"
.SH NAME
fedfs-create-junction \- send a FEDFS_CREATE_JUNCTION ADMIN protocol request
.SH SYNOPSIS
.B fedfs-create-junction
.RB [ \-?d ]
.RB [ \-n
.IR nettype ]
.RB [ \-h
.IR hostname ]
.RB [ \-l
.IR nsdbname ]
.RB [ \-r
.IR nsdbport ]
.I path
.I fsn-uuid
.SH INTRODUCTION
RFC 5716 introduces the Federated File System (FedFS, for short).
FedFS is an extensible standardized mechanism
by which system administrators construct
a coherent namespace across multiple file servers using
.IR "file system referrals" .
For further details, see
.BR fedfs (7).
.P
FedFS-enabled file servers allow remote administrative access via an
authenticated RPC protocol known as the
.IR "FedFS ADMIN protocol" .
Using this protocol, FedFS administrators manage
FedFS junctions and NSDB connection parameter information
on remote FedFS-enabled file servers.
.SH DESCRIPTION
The
.BR fedfs-create-junction (8)
command is part of a collection of low-level single-use programs
that is intended for testing the FedFS ADMIN protocol or for use in scripts.
It sends a single FEDFS_CREATE_JUNCTION request to a remote
FedFS ADMIN protocol service.
.P
The FEDFS_CREATE_JUNCTION request creates a FedFS junction
in a local file system on a remote file server.
The contents of a FedFS junction are an FSN UUID and an NSDB name and port.
.P
The
.BR fedfs-create-junction (8)
command takes two positional parameters which specify
the pathname on the remote server of the new junction, and the FSN UUID.
This pathname is relative to the root
of the local file system on the remote server.
Required NSDB information can be inferred
from the command's environment or specified on the command line.
The meaning of these arguments is described in more detail in
.BR fedfs (7).
.P
The FEDFS_CREATE_JUNCTION request does not create an FSN record.
To create an FSN record, use the
.BR nsdb-create-fsn (8)
command.
Resolving a junction that contains an FSN UUID without
a matching FSN record on the NSDB results in an error on the file server.
.SH OPTIONS
.IP "\fB\-d, \-\-debug"
Enables debugging messages during operation.
.IP "\fB\-?, \-\-help"
Displays
.BR fedfs-create-junction (8)
version information and a usage message on
.IR stderr .
.IP "\fB\-h, \-\-hostname=\fIhostname\fP"
Specifies the hostname of a remote FedFS ADMIN service.
If this option is not specified, the default value is
.BR localhost .
.IP "\fB\-n, \-\-nettype=\fInettype\fP"
Specifies the transport to use when contacting the remote FedFS ADMIN service.
Typically the
.I nettype
is one of
.B tcp
or
.BR udp .
If this option is not specified, the default value is
.BR netpath .
See
.BR rpc (3t)
for details.
.IP "\fB\-l, \-\-nsdbname=\fINSDB-hostname\fP"
Specifies the hostname of the NSDB to insert into the new FedFS junction.
If this option is not specified,
the value of the FEDFS_NSDB_HOST environment variable is consulted.
If the variable is not set and the
.B \--nsdbname
option is not specified, the
.BR fedfs-create-junction (8)
command fails.
.IP "\fB\-r, \-\-nsdbport=\fINSDB-port\fP"
Specifies the IP port of the NSDB to insert into the new FedFS junction.
If this option is not specified,
the value of the FEDFS_NSDB_PORT environment variable is consulted.
The default value if the variable is not set is 389.
.SH EXIT CODES
The remote administration service returns a value that reflects the
success of the requested operation.
.TP
.B FEDFS_OK
The operation succeeded.
.TP
.B FEDFS_ERR_ACCESS
The caller does not have permission to perform the requested operation
on the remote server.
.TP
.B FEDFS_ERR_BADCHAR
The pathname contains a character which is not
supported by the remote server.
.TP
.B FEDFS_ERR_BADNAME
A component of the pathname consisted of valid UTF-8 characters
supported by the remote server,
but the name is not supported by the remote server
as a valid name for the current operation.
.TP
.B FEDFS_ERR_NAMETOOLONG
The length of the pathname exceeds the remote server’s implementation limit.
.TP
.B FEDFS_ERR_LOOP
Too many symbolic links were encountered in resolving the pathname.
.TP
.B FEDFS_ERR_BADXDR
The remote server encountered an XDR decoding error while
processing the request.
.TP
.B FEDFS_ERR_EXIST
The specified junction already exists.
.TP
.B FEDFS_ERR_INVAL
One of the arguments was not valid.
.TP
.B FEDFS_ERR_IO
A hard error occurred on the remote server.
.TP
.B FEDFS_ERR_NOSPC
The requested operation would have caused the remote
server’s filesystem to exceed some limit (for example, if there is
a fixed number of junctions per fileset or per server).
.TP
.B FEDFS_ERR_NOTJUNCT
The pathname does not end in a junction.
.TP
.B FEDFS_ERR_NOTLOCAL
The pathname contains a junction in any position other than the last component.
.TP
.B FEDFS_ERR_PERM
The operation was not allowed because the caller is
either not a privileged user or not the owner of an object that
would be modified by the operation.
.TP
.B FEDFS_ERR_ROFS
A modifying operation was attempted on a read-only filesystem.
.TP
.B FEDFS_ERR_SVRFAULT
An unanticipated non-protocol error occurred on the remote server.
.TP
.B FEDFS_ERR_NSDB_PARAMS
The remote server does not have any connection
parameters on record for the specified NSDB.
.TP
.B FEDFS_ERR_PATH_TYPE_UNSUPP
The remote server does not support the specified FedFsPathType value.
.TP
.B FEDFS_ERR_NOTSUPP
The remote server does not support the specified procedure.
.SH EXAMPLES
Suppose you are the FedFS administrator of the
.I example.net
FedFS domain and that your domain's NSDB hostname is
.IR nsdb.example.net .
To create a new FedFS junction on the file server
.IR fs.example.net ,
use:
.RS
.sp
$ fedfs-create-junction -h fs.example.net -l nsdb.example.net \\
	/export/junction1 `uuidgen -t`
.sp
.RE
In this example, a new FSN UUID is created on the spot.
It can be read back from the remote server using the
.BR fedfs-lookup-junction (8)
command, and added to the NSDB using the
.BR nsdb-create-fsn (8)
command.
.SH SECURITY
RPCSEC GSSAPI authentication has not yet been implemented for this command.
.SH "SEE ALSO"
.BR fedfs (7),
.BR rpc.fedfsd (8),
.BR fedfs-lookup-junction (8),
.BR nsdb-create-fsn (8),
.BR rpc (3t)
.sp
RFC 5716 for FedFS requirements and overview
.SH COLOPHON
This page is part of the fedfs-utils package.
A description of the project and information about reporting bugs
can be found at
.IR http://wiki.linux-nfs.org/wiki/index.php/FedFsUtilsProject .
.SH "AUTHOR"
Chuck Lever <chuck.lever@oracle.com>


src/src/fedfsc/fedfs-create-junction.c
/**
 * @file src/fedfsc/fedfs-create-junction.c
 * @brief Send a FEDFS_CREATE_JUNCTION RPC to a FedFS ADMIN server
 */

/*
 * Copyright 2010, 2013 Oracle.  All rights reserved.
 *
 * This file is part of fedfs-utils.
 *
 * fedfs-utils is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2.0 as
 * published by the Free Software Foundation.
 *
 * fedfs-utils is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License version 2.0 for more details.
 *
 * You should have received a copy of the GNU General Public License
 * version 2.0 along with fedfs-utils.  If not, see:
 *
 *	http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt
 */

#include <sys/types.h>
#include <sys/stat.h>

#include <stdbool.h>
#include <stdlib.h>
#include <fcntl.h>
#include <unistd.h>
#include <errno.h>
#include <getopt.h>
#include <locale.h>

#include "fedfs.h"
#include "fedfs_admin.h"
#include "admin.h"
#include "nsdb.h"
#include "junction.h"
#include "xlog.h"
#include "gpl-boiler.h"

/**
 * Short form command line options
 */
static const char fedfs_create_junction_opts[] = "?dh:l:n:r:s:";

/**
 * Long form command line options
 */
static const struct option fedfs_create_junction_longopts[] = {
	{ "debug", 0, NULL, 'd', },
	{ "help", 0, NULL, '?', },
	{ "hostname", 1, NULL, 'h', },
	{ "nsdbname", 1, NULL, 'l', },
	{ "nettype", 1, NULL, 'n', },
	{ "nsdbport", 1, NULL, 'r', },
	{ "security", 1, NULL, 's', },
	{ NULL, 0, NULL, 0, },
};

/**
 * Display program synopsis
 *
 * @param progname NUL-terminated C string containing name of program
 * @return program exit status
 */
static int
fedfs_create_junction_usage(const char *progname)
{
	fprintf(stderr, "\n%s version " VERSION "\n", progname);
	fprintf(stderr, "Usage: %s [-d] [-n nettype] [-h hostname] "
			"[-l nsdbname] [-r nsdbport] "
			"pathname fsn-uuid\n\n",
			progname);

	fprintf(stderr, "\t-?, --help           Print this help\n");
	fprintf(stderr, "\t-d, --debug          Enable debug messages\n");
	fprintf(stderr, "\t-n, --nettype        RPC transport (default: 'netpath')\n");
	fprintf(stderr, "\t-h, --hostname       ADMIN server hostname (default: 'localhost')\n");
	fprintf(stderr, "\t-l, --nsdbname       NSDB hostname to set\n");
	fprintf(stderr, "\t-r, --nsdbport       NSDB port to set\n");
	fprintf(stderr, "\t-s, --security       RPC security level\n");

	fprintf(stderr, "%s", fedfs_gpl_boilerplate);

	return EXIT_FAILURE;
}

/**
 * Create a junction on a remote fileserver
 *
 * @param host an initialized and opened admin_t
 * @param path_array an array of NUL-terminated C strings containing pathname components
 * @param fsn FileSet Name to plant in new junction
 * @return program exit status
 */
static int
fedfs_create_junction_try(admin_t host, char * const *path_array,
		struct admin_fsn *fsn)
{
	int status, err;

	status = EXIT_FAILURE;
	err = admin_create_junction(host, path_array, fsn);
	switch (err) {
	case 0:
		break;
	case EACCES:
		xlog(L_ERROR, "%s: access denied", admin_hostname(host));
		xlog(D_GENERAL, "%s",
			admin_perror(host, admin_hostname(host)));
		goto out;
	case EIO:
		xlog(L_ERROR, "%s",
			admin_perror(host, admin_hostname(host)));
		goto out;
	default:
		xlog(L_ERROR, "ADMIN client: %s", strerror(err));
		goto out;
	}

	switch (admin_status(host)) {
	case FEDFS_OK:
		printf("Junction created successfully\n");
		status = EXIT_SUCCESS;
		break;
	case FEDFS_ERR_NSDB_PARAMS:
		printf("No connection parameters found\n");
		break;
	default:
		nsdb_print_fedfsstatus(admin_status(host));
	}

out:
	return status;
}

/**
 * Create a junction on a remote fileserver
 *
 * @param hostname NUL-terminated UTF-8 string containing ADMIN server's hostname
 * @param nettype NUL-terminated C string containing nettype to use for connection
 * @param security NUL-terminated C string containing RPC security mode
 * @param path_array an array of NUL-terminated C strings containing pathname components
 * @param fsn FileSet Name to plant in new junction
 * @return program exit status
 */
static int
fedfs_create_junction_host(const char *hostname, const char *nettype,
		const char *security, char * const *path_array,
		struct admin_fsn *fsn)
{
	admin_t host;
	int status;

	status = EXIT_FAILURE;
	switch (admin_create(hostname, nettype, security, &host)) {
	case 0:
		status = fedfs_create_junction_try(host, path_array, fsn);
		admin_release(host);
		break;
	case EINVAL:
		xlog(L_ERROR, "Invalid command line parameter");
		break;
	case EACCES:
		xlog(L_ERROR, "Failed to authenticate server");
		break;
	case EKEYEXPIRED:
		xlog(L_ERROR, "User credentials not found");
		break;
	default:
		xlog(L_ERROR, "%s",
			admin_open_perror(hostname));
	}

	return status;
}

/**
 * Program entry point
 *
 * @param argc count of command line arguments
 * @param argv array of NUL-terminated C strings containing command line arguments
 * @return program exit status
 */
int
main(int argc, char **argv)
{
	char *progname, *hostname, *nettype, *security, *path;
	struct admin_nsdb *nsdb;
	struct admin_fsn fsn;
	FedFsStatus retval;
	char **path_array;
	int arg, status;

	(void)setlocale(LC_ALL, "");
	(void)umask(S_IRWXO);

	/* Set the basename */
	if ((progname = strrchr(argv[0], '/')) != NULL)
		progname++;
	else
		progname = argv[0];

	/* For the libraries */
	xlog_stderr(1);
	xlog_syslog(0);
	xlog_open(progname);

	nsdb = &fsn.af_nsdb;
	nsdb_env((char **)&nsdb->an_hostname, &nsdb->an_port, NULL, NULL);

	hostname = "localhost";
	nettype = "netpath";
	security = "unix";
	while ((arg = getopt_long(argc, argv, fedfs_create_junction_opts,
			fedfs_create_junction_longopts, NULL)) != -1) {
		switch (arg) {
		case 'd':
			xlog_config(D_ALL, 1);
			break;
		case 'h':
			hostname = optarg;
			break;
		case 'l':
			if (!nsdb_is_hostname_utf8(optarg)) {
				fprintf(stderr, "NSDB name %s is "
					"not a UTF-8 hostname\n", optarg);
				return fedfs_create_junction_usage(progname);
			}
			nsdb->an_hostname = optarg;
			break;
		case 'n':
			nettype = optarg;
			break;
		case 'r':
			if (!nsdb_parse_port_string(optarg, &nsdb->an_port)) {
				fprintf(stderr, "Bad port number: '%s'\n",
					optarg);
				return fedfs_create_junction_usage(progname);
			}
			break;
		case 's':
			security = optarg;
			break;
		default:
			fprintf(stderr, "Invalid command line "
				"argument: %c\n", (char)arg);
		case '?':
			return fedfs_create_junction_usage(progname);
		}
	}
	if (argc == optind + 2) {
		path = argv[optind];
		fsn.af_uuid = argv[optind + 1];
	} else {
		fprintf(stderr, "Ambiguous positional parameters\n");
		return fedfs_create_junction_usage(progname);
	}
	if (nsdb->an_hostname == NULL) {
		fprintf(stderr, "No NSDB hostname was specified\n");
		return fedfs_create_junction_usage(progname);
	}

	retval = nsdb_posix_to_path_array(path, &path_array);
	if (retval != FEDFS_OK) {
		fprintf(stderr, "Failed to encode pathname: %s",
			nsdb_display_fedfsstatus(retval));
		return EXIT_FAILURE;
	}

	status = fedfs_create_junction_host(hostname, nettype, security,
						path_array, &fsn);

	nsdb_free_string_array(path_array);
	return status;
}
